### ISSUE TITLE
Key metadata is stored unencrypted

### ISSUE THREAD

**Issue by nioncode** [restic/restic#2128]
Output of `restic version`
--------------------------
restic 0.9.2 compiled with go1.11 on linux/amd64

How did you run restic exactly?
-------------------------------
restic -r /repo init

What backend/server/service did you use to store the repository?
----------------------------------------------------------------
local

Expected behavior
-----------------

Keys should contain no plaintext information (apart from salt, iteration, ...).

Actual behavior
---------------
Keys contain host name, username, and time of creation in plaintext.

Do you have any idea what may have caused this?
-----------------------------------------------
Restic stores key metadata unencrypted.

Do you have an idea how to solve the issue?
-------------------------------------------
Store key metadata encrypted. We could update the key format by adding e.g. a `hostname_enc` field and blanking the `hostname` field for new repos. Then old clients would show empty metadata until they are updated. We should also migrate existing keys to remove potentially old plaintext metadata from existing repos.

**Comment by fd0** [comment]
We'll probably just remove the fields from the key metadata, that should be sufficient.

**Comment by nioncode** [comment]
How are keys identified then, if all their metadata gets removed? Or do you just want to remove the plaintext fields and introduce encrypted ones?

**Comment by fd0** [comment]
Yes, through the key ID (the filename prefix).

**Comment by cdhowie** [comment]
Perhaps have this be optional. We use a key per backup host and it's nice to be able to quickly see which key belongs to which host.  Without this data in restic, we'd have to maintain the data somewhere else (company wiki, for example) -- not the worst thing, but why fix what isn't broken?

**Comment by nioncode** [comment]
I agree that it's nice to have, but this leaks metadata unencrypted. That's why I proposed to have them stored encrypted, so that each key that can decrypt the repo can still read all metadata like it is now, but the not trusted backup location can not gather the metadata in plaintext. This way, there would be no user facing changes at all.

**Comment by cdhowie** [comment]
> there would be no user facing changes at all.

That depends if the repository is automatically migrated to store the data encrypted.  If so, the key IDs will necessarily change and, depending on syncing processes/restrictions, the user may need to be aware that this happened.

**Comment by monkeyhybrid** [comment]
If this metadata is to be removed, could we at least have an optional 'tag' attribute for each key that can contain anything we wish? Make the tag empty by default for those concerned with leaking metadata, but allow custom text that could be useful for key management, for those that are not.

**Comment by ducalex** [comment]
When stripping the metadata manually I get the following:

````
Load(<key/5d081d8a51>, 0, 0) returned error, retrying after 552.330144ms: loadAll(<key/5d081d8a51>): invalid data returned
Load(<key/5d081d8a51>, 0, 0) operation successful after 1 retries
````

The error is expected, as the id no longer matches the content, but the successful retry is a (nice) surprise to me. I can't find in the code where it loosens the check and accept the content anyway. Can someone point me to where that happens?

**Comment by MichaelEischer** [comment]
@ducalex The filename of a file stored in the repository *must* always match the SHA256 hash of its content. Thus after tampering with a key, you have to update its filename. Removing integrity checks isn't a good idea.

> Load(<key/5d081d8a51>, 0, 0) operation successful after 1 retries

Hmm, that error message isn't particularly helpful...

**Comment by ducalex** [comment]
> @ducalex The filename of a file stored in the repository _must_ always match the SHA256 hash of its content. Thus after tampering with a key, you have to update its filename. Removing integrity checks isn't a good idea.

That is what I would expect, but why does it eventually succeed anyway?

Example:

````
$ mkdir temp
$ restic -rtemp init
enter password for new repository:
enter password again:
created restic repository 5a13f78dee at .
$ vi temp/keys/6de9da466bb841d144529a04690a7d0ca3aa1caa3fa0c8aa8416f67cae34f521 # strip metadata
$ touch test
$ restic -rtemp backup test
enter password for repository:
Load(<key/6de9da466b>, 0, 0) returned error, retrying after 552.330144ms: loadAll(<key/6de9da466b>): invalid data returned
Load(<key/6de9da466b>, 0, 0) operation successful after 1 retries
Load(<key/6de9da466b>, 0, 0) operation successful after 1 retries
Load(<key/6de9da466b>, 0, 0) returned error, retrying after 582.280027ms: loadAll(<key/6de9da466b>): invalid data returned
Load(<key/6de9da466b>, 0, 0) operation successful after 1 retries
repository 5a13f78d opened (version 2, compression level auto)
no parent snapshot found, will read all files

Files:           1 new,     0 changed,     0 unmodified
Dirs:            0 new,     0 changed,     0 unmodified
Added to the repository: 448 B (488 B stored)
processed 1 files, 155 B in 0:00
snapshot bcb08633 saved
````

**Comment by MichaelEischer** [comment]
> That is what I would expect, but why does it eventually succeed anyway?

restic retries downloading a damaged file at most once. Afterwards the data is passed on to whatever component requested it, independent of whether that data is correct or not. Depending on what that component does, this will either be successful (key is usable as the JSON can still be decoded) or fail (e.g. what trying to decrypt damaged data).

The reason why invalid data might be passed on to the next component is that this can be useful e.g. when trying to repair or analyze a damaged repository.

**Comment by mozlima** [comment]
Any news? 
This leak hostname and username
I came here after searching in the repository.
`grep -rE 'myuser|myhost|otherinfos|user|host' /repos*`

I hope this was encrypted  binary file.

**Comment by sergeevabc** [comment]
Aggrrhh! Waiting for this to be resolved since [2017-06-19](https://github.com/restic/restic/issues/628#issuecomment-309388855).

**Comment by rawtaz** [comment]
I take it such a change would require a new repository version?

**Comment by MichaelEischer** [comment]
The keys are decoded before the repository format version can even be read from the `config` file. So, whatever change is made should still allow old restic version to still read those keys. But not being able to read the metadata would be fine as it's currently only used by `restic key list`. In short, we can change this independently of the repository format.

As nobody has provided a use case for the name+host fields, let's just drop them. Then add a new label field, which should be encrypted using the masterkey. As there have been complaints about the creation timestamp leaking the time zone, let's just switch to storing it as UTC. The creation time of the file is visible in the file metadata anyways, so there's no point in additionally encrypting the timestamp.

**Comment by denvolok** [comment]
I noted that the `restic key add` command allows specifying `--host` and `--user` params, so you can create a new key for your repo and mock those values.
As for timezone leak in the creation date timestamp, `TZ=UTC` env var can be used.

After new key added, you can remove the initial key.  Perhaps, if your current key is shared (or mirrored in multiple locations) it will not be an option.

Here is the full command I used:
```sh
export TZ=UTC && restic -r ~/repo key add --host host --user user
```

Not sure if I missed something (as I just started exploring Restic).
